### DISCUSSION TITLE
How can I uninstall or reinstall PDM?

### DISCUSSION THREAD

**Discussion by lazarillo** [pdm-project/pdm#713] [General]
I updated my version of Python -- I use `brew` and expect / assume that I can update many packages and the links will update accordingly.

Now `pdm` is associated with a version that no longer exists.  However, due to a deep nest of soft links and pointers that `pdm` has to multiple files representing Python, there is no easy way to just alter a line of code somewhere or redirect a soft link.

I would therefore like to uninstall `pdm` and reinstall it.  (This time I will install via `brew` to ensure links are followed properly, despite the recommendation the the PDM documentation page not to do this.)  There is no `uninstall` command, at least not that I can see.  But since `pdm` is broken, a command like `pdm uninstall` would not help anyway.

I did a `find` command, and there is a giant mess of files created.  The snippet below is just a very small portion of them:

```bash
./System/Volumes/Data/Users/mikewilliamson/Library/Application Support/Code/User/workspaceStorage/9b36b749397a0ddb6f3f86c0fb94b440/redhat.java/jdt_ws/protobufs_8db2cab0/bin/pdm.lock
./System/Volumes/Data/Users/mikewilliamson/Library/Application Support/Code/User/workspaceStorage/9b36b749397a0ddb6f3f86c0fb94b440/redhat.java/jdt_ws/protobufs_8db2cab0/bin/.pdm.toml
./System/Volumes/Data/Users/mikewilliamson/Library/Application Support/pdm
./System/Volumes/Data/Users/mikewilliamson/Library/Application Support/pdm/venv/bin/pdm
./System/Volumes/Data/Users/mikewilliamson/Library/Application Support/pdm/venv/lib/python3.9/site-packages/pdm-1.9.0.dist-info
./System/Volumes/Data/Users/mikewilliamson/Library/Application Support/pdm/venv/lib/python3.9/site-packages/pdm_pep517-0.8.3.dist-info
./System/Volumes/Data/Users/mikewilliamson/Library/Application Support/pdm/venv/lib/python3.9/site-packages/pdm
./System/Volumes/Data/Users/mikewilliamson/Library/Application Support/pdm/venv/lib/python3.9/site-packages/pdm/cli/completions/pdm.bash
./System/Volumes/Data/Users/mikewilliamson/Library/Application Support/pdm/venv/lib/python3.9/site-packages/pdm/cli/completions/pdm.fish
./System/Volumes/Data/Users/mikewilliamson/Library/Application Support/pdm/venv/lib/python3.9/site-packages/pdm/cli/completions/pdm.ps1
./System/Volumes/Data/Users/mikewilliamson/Library/Application Support/pdm/venv/lib/python3.9/site-packages/pdm/cli/completions/pdm.zsh
./System/Volumes/Data/Users/mikewilliamson/Library/Python/3.9/bin/pdm
./System/Volumes/Data/Users/mikewilliamson/Library/Caches/com.apple.python/Users/mikewilliamson/Library/Application Support/pdm
./System/Volumes/Data/Users/mikewilliamson/Library/Caches/com.apple.python/Users/mikewilliamson/Library/Application Support/pdm/venv/lib/python3.9/site-packages/pdm
./System/Volumes/Data/Users/mikewilliamson/Library/Caches/pdm
./System/Volumes/Data/Users/mikewilliamson/.oh-my-zsh/custom/plugins/pdm
./System/Volumes/Data/Users/mikewilliamson/.oh-my-zsh/custom/plugins/pdm/_pdm
./Users/mikewilliamson/Library/Application Support/pdm
./Users/mikewilliamson/Library/Application Support/pdm/venv/bin/pdm
./Users/mikewilliamson/Library/Application Support/pdm/venv/lib/python3.9/site-packages/pdm-1.9.0.dist-info
./Users/mikewilliamson/Library/Application Support/pdm/venv/lib/python3.9/site-packages/pdm_pep517-0.8.3.dist-info
./Users/mikewilliamson/Library/Application Support/pdm/venv/lib/python3.9/site-packages/pdm
./Users/mikewilliamson/Library/Application Support/pdm/venv/lib/python3.9/site-packages/pdm/cli/completions/pdm.bash
./Users/mikewilliamson/Library/Application Support/pdm/venv/lib/python3.9/site-packages/pdm/cli/completions/pdm.fish
./Users/mikewilliamson/Library/Application Support/pdm/venv/lib/python3.9/site-packages/pdm/cli/completions/pdm.ps1
./Users/mikewilliamson/Library/Application Support/pdm/venv/lib/python3.9/site-packages/pdm/cli/completions/pdm.zsh
./Users/mikewilliamson/Library/Python/3.9/bin/pdm
./Users/mikewilliamson/Library/Caches/com.apple.python/Users/mikewilliamson/Library/Application Support/pdm
./Users/mikewilliamson/Library/Caches/com.apple.python/Users/mikewilliamson/Library/Application Support/pdm/venv/lib/python3.9/site-packages/pdm
./Users/mikewilliamson/Library/Caches/pdm
```

There are truly hundreds more files, links, etc which are not to any packages I am managing with PDM, but just PDM itself.  How can I uninstall and reinstall it?

**Comment by lazarillo** [comment]
Sorry, I figured it out.  I can re-run the `curl` command that installs the script, but passing the `--remove` flag or setting the `PDM_REMOVE` env var.

**Reply by lazarillo** [reply]
This is in the documentation, I just missed it.

**Comment by frostming** [comment]
Just avoid using brew python to install tools like pdm. The way brew uses to manage Python version makes it a pain when Python gets upgraded.

From now on, you should not use brew python to install tools, if there are, re-install them. However, you can use the tool's homebrew version if is provided. `brew install pdm` and `brew upgrade` can upgrade it smoothly afterward.

**Reply by lazarillo** [reply]
Actually, I found I had the opposite problem:  Brew install of PDM was better able to match the Python version I would want to use than the curl install of PDM.

I cannot remember the details now, but I specifically installed PDM first via curl, then uninstalled as described above so that I could install with brew.  Since the new install via brew, I have had no problems.

**Reply by lazarillo** [reply]
This is not a complaint, by the way:  I think that PDM is awesome, and I love using it now.  I also recognize that brew does make links complicated.  However, I think that brew has a fairly complete and robust linking system that captures edge cases well, and I'm not sure the install via curl is doing this as well.

Again, I am ignorant of the details, just saying that brew actually works quite well, IMHO.

**Reply by frostming** [reply]
I am not against brew install of PDM, it is listed as one of the recommendations. I am against installing tools using the `pip` shipped by a brew-installed Python. Note the difference?

**Reply by lazarillo** [reply]
Yes, I understand.  Thanks!  :)
